implement :echomsg!

justinmk opened the pull request:
Followup of my comments here: https://github.com/neovim/neovim/issues/1524#issuecomment-64085698 , https://github.com/neovim/neovim/issues/1524#issuecomment-64090515 , https://github.com/neovim/neovim/issues/1524#issuecomment-64097010

Closes #2063
Closes #5103
ref https://github.com/neovim/neovim/issues/2308 https://github.com/neovim/neovim/issues/1755 https://github.com/neovim/neovim/issues/1029
ref https://github.com/neovim/neovim/issues/7291
ref https://github.com/neovim/neovim/issues/7290
ref https://github.com/neovim/neovim/issues/7326

This PR implements "bang" variants of `:echomsg` ~~and `:echoerr`~~ (**edit:** not going to implement `:echoerr!`; the semantics of `:echoerr` differ greatly from `:echomsg`, and have implications for try/catch--it wouldn't make sense for :echoerr to be "passive"). Documented as follows:

```
+                                                     *:echom!* *:echomsg!* 
+           |:echomsg|  followed by [!] creates a
+           "transient" message, which will never cause a screen
+           scroll or a |press-enter| prompt, and is _always_
+           added to the |:messages| history (even if |:silent| or
+           |:map-<silent>| are used). This is useful for leaving
+           passive diagnostic messages without disturbing
+           usability.
```

Besides solving a general usability issue (avoiding the "Press ENTER" prompt for non-critical messages), this enables us to be much more consistent and less hesitant about informing the user of possibly-useful context-sensitive discoveries while bootstrapping official plugins, providers (python, clipboard), and general sanity checks (VIMRUNTIME).

Before writing tests, I would like feedback on the general approach. I usually advocate against extending vimscript, but after much thought I could not find a reasonable medium-term solution. Some ideas that I rejected:
- extend `histadd()` to allow adding to the `:messages` history.
  - This is icky becaus then we have to also implement `histdel()` _or_ introduce a glaring incongruity.
- expose `:messages` list as a read-write `v:messages` List.
  - Same problem as above: giving plugins the ability to delete from `:messages` is a bad idea.
- create an API-only function for adding to history
  - This is fine, but I suspect that users will really appreciate being able to easily avoid the "Press ENTER" prompt.
- add a new flag to `'shortmess'`
  - This is still an option that I think is worth exploring, in addition to providing `:echomsg!` and `:echoerr!`. But it does not solve the use cases mentioned above (providers, sanity checks).

---

**`:echomsg!` spec**
- [x] writes to history even with `:silent[!]` and `<silent>`
- [x] no duplicates in history
- [x] long message(s) emmitted from a keymap-triggered script are truncated
  - [x] if 'cmdheight' is N, message fills up to N lines, but no more.
- [x] echom! emmitted from command line :call Foo() invocation are truncated?
- [x] :echom! at user-entered command line should act the same as if from a script
- [x] test regressions: :exe, :echom, :echoerr, :echo, :echohl, printf()
- [ ] behavior with laststatus=0
- [x] test with ":set noshowcmd"
- [x] test with `set shortmess-=T`
- [x] `v:statusmsg` is set only by `:echom[!]`
- [x] `v:errmsg` is set only by `:echoerr`
- [ ] CTRL-G should overwrite like `:echo`
- [ ] If `cmdheight=2` but there is a 1-line message displayed, `:echom!` should not clear the existing message unless it needs more than 1 line.

**"+N :messages" spec**
- [ ] test overflow (very large number of notifications)
- [ ] displays count after event loop
- [ ] count resets after next user action
- [ ] showcmd blanks out notification area
- [ ] appears after all messages posted (until next user command)
- [ ] appears only if 2 or more non-duplicate messages since last batch
- [ ] appears in correct location

justinmk committed:
messages.c: mark non-null; convert to bool

rename:
  delete_first_msg  => msg_delete_first
  add_msg_hist      => msg_hist_add

justinmk committed:
ex_cmds.c: document print_line_no_prefix()

justinmk committed:
remove USE_MCH_ERRMSG (restore later in some fashion)

This is used to collect error messages while the GUI is bootstrapping.
We may need some form of this logic for the TUI/GUI clients, but instead
of defining mch_errmsg() as fprintf(), etc., we should havea single
implementation that raises API events if appropriate.

justinmk committed:
IWYU: move message-related symbols from vim.h to message.h

justinmk committed:
vim.h: remove outdated comment. TODO: define mch_errmsg, display_errors, mch_msg, directly instead of using #define.

justinmk committed:
move PERROR to message.h; remove outdated comment.

justinmk committed:
message.h: increase SHOWCMD_COLS to make room for  "+NN :messages".

justinmk committed:
doc: :echomsg!, :echoerr!

justinmk committed:
ex_messages: remove "Messages maintainer" message

justinmk committed:
message.h: increas MAX_MSG_HIST_LEN to 400 (was 200)

justinmk committed:
messages: implement :echomsg!, :echoerr!

tarruda commented:
> Before writing tests, I would like feedback on the general approach. I usually advocate against extending vimscript, but after much thought I could not find a reasonable medium-term solution. Some ideas that I rejected:

Why not simply use the screen verification facility for testing this? [Here's](https://github.com/neovim/neovim/blob/master/test/functional/ui/screen_basic_spec.lua#L203-L221) an example

justinmk commented:
@tarruda that is my plan, I just didn't want to continue if there were objections to this PR.

tarruda commented:
> @tarruda that is my plan, I just didn't want to continue if there were objections to this PR.

No objections, have you also proposed this on vim-dev?

justinmk commented:
> have you also proposed this on vim-dev?

Wasn't planning on it, but I will.

Shougo commented:
:+1:

aktau commented on src/nvim/memory.c:
```diff
@@ -596,9 +596,7 @@ void free_all_mem(void)
-  for (;; )
-    if (delete_first_msg() == FAIL)
-      break;
+  while(msg_delete_first()) {;}
```
As far as I've seen this is the first instance of the `{;}` pattern, I though we were using just `{}`.

justinmk replied:
Will fix

justinmk committed:
test: echomsg_spec.lua [WIP]

Pyrohh commented:
@justinmk I see that you're including some changes to message.h in this PR, perhaps you could pull https://github.com/Pyrohh/neovim/commit/374d8200bfc2690232917228590862371f5dff13 in as well? I can just submit a PR if it's too far reaching.

justinmk committed:
test: echom, echom!

justinmk committed:
WIP

justinmk committed:
test: 'shortmess' without "T" (truncate) flag

justinmk committed:
messages: specify MSG_HIST at set_keep_msg() calls

msg_attr_keep() adds keep_msg to the :messages history if the keep_msg
does not match the most recent history item. But msg_passive() sets
a _truncated_ keep_msg, which obviously will not match. We do not want
the truncated message in the :messages history (the full message is
already there), so we need to avoid adding it to the history. Instead of
adding another global boolean, send MSG_HIST explicitly at each
set_keep_msg() call.

justinmk committed:
test: Screen:expect: handle incorrect number of rows

justinmk committed:
msg_passive(): do not display if :silent

justinmk committed:
test: :echoerr, :echom! in a {function, mapping}

justinmk committed:
test: :echom! adds the non-truncated message to :messages

Also test for :echomsg, :echoerr regressions.

justinmk committed:
test: :echom! adds to :messages even if :silent

Also test for :echom regression.

justinmk commented on test/functional/ex_cmds/echomsg_spec.lua:
```diff
@@ -0,0 +1,494 @@
+  end)
+
+
+  local function echomsg_bang_tests(cmd_under_test)
```
This test group is executed twice for `shortmess-=T` and `shortmess+=T`.

justinmk commented on test/functional/ex_cmds/echomsg_spec.lua:
```diff
@@ -0,0 +1,494 @@
+    end)
+  end)
+
+  describe(':echoerr (highlight)', function()
```
These are just regression tests.

justinmk commented on test/functional/ex_cmds/echomsg_spec.lua:
```diff
@@ -0,0 +1,494 @@
+    screen:detach()
+  end)
+
+  describe(':echomsg', function()
```
Regression tests for `:echomsg`.

justinmk commented on test/functional/ui/screen.lua:
```diff
@@ -202,9 +202,11 @@ function Screen:expect(expected, attr_ids, attr_ignore)
-        return 'Row '..tostring(i)..' didnt match.\nExpected: "'..
-               expected_row..'"\nActual:   "'..actual_row..'"'
+      if not expected_row then
+        return 'Screen has more rows than expected.'
```
During test development, if a `screen:expect()` call has too few lines, `expected_row` is nil. So this shows a better error message in that case.

Perhaps we should also do a sanity check for too *many* expected rows.

justinmk commented on test/functional/ui/screen.lua:
```diff
@@ -202,9 +202,11 @@ function Screen:expect(expected, attr_ids, attr_ignore)
+        return 'Screen has more rows than expected.'
+      elseif expected_row ~= actual_row then
+        return 'row '..tostring(i)..':\n  expected: "'..
+               expected_row..'"\n  actual:   "'..actual_row..'"'
```
Small layout improvement (IMO) for test failure message.

justinmk committed:
test: v:statusmsg, v:errmsg

justinmk commented:
@tarruda After rebasing, I have an unexpected [test](https://github.com/neovim/neovim/pull/1802/files#diff-e0a79ed788f67783cce273283c007e8eR521) failure in this PR:

```
row 1:
  expected: "^                                                    "
  actual:   "                                                     "
```

If I change the expected `^` in row 1 to a `~`, then the test fails with the converse:

```
expected: "~                                                    "
actual:   "^                                                    "
```

All tests were passing before the rebasing on TUI.

tarruda commented:
@justinmk I tested this branch, and it seems you are experiencing a race condition. This patch should fix it:

```diff
diff --git a/test/functional/ex_cmds/echomsg_spec.lua b/test/functional/ex_cmds/echomsg_spec.lua
index 0cea285..cf05cba 100644
--- a/test/functional/ex_cmds/echomsg_spec.lua
+++ b/test/functional/ex_cmds/echomsg_spec.lua
@@ -523,6 +523,7 @@ describe('echomsg', function()
       _h.nvim('set_option', 'cmdheight', 2)
       execute('nnoremap <silent> foo :silent '..cmd_under_test..' "'..fullmsg..'"<cr>')
       _h.feed('foo')
+      _h.wait()
       screen:expect([[
       ^                                                    |
       ~
```

`wait()` It is simply a way of saying: "Wait until all keys I've sent are fully processed". This is required because the `_h.nvim('set_option', 'cmdheight', 2)` may set the screen to the state you expect it to be later, and the first redraw notification will succeed, only to be followed by another that fails(sent while `foo` is being processed)

justinmk commented:
@tarruda Thanks for looking into that. I was being paranoid so I didn't want to just add a `wait()` to make it go away, but now that you explain it I suppose there is no reason to expect any particular sequence of redraws until all keys are processed.

marlun commented:
Been some time since there's been an update here, just curious if this is still being worked on? I'm getting `No neovim module found for Python 2.7.6. Try installing it with 'pip install neovim' or see ':help nvim-python'.` everytime I open vim probably because vim-plug is running `has('python')` but since it works fince without python and I'm not using any other plugin that requires python it would be nice if I didn't have to hit *ENTER* everytime I open nvim.

justinmk commented:
Yes, I need to get this out the door.

justinmk committed:
test: WIP

justinmk committed:
WIP: test cmdheight=4

justinmk committed:
remove cursor_off

see c546875daf36936b9a6c0886a71c9edd1fdae6db

justinmk committed:
rew: fix tests

fmoralesc commented:
Any progress in this? #2769 would benefit a lot from this, for example.

justinmk commented:
@fmoralesc I'm thinking the OS X builds are higher priority, but then this was my next priority.

fmoralesc commented:
Indeed! Carry on... I was just reminded of this when reviewing #2769, the messages there should probably use something like `silent echomsg!`.

wsdjeg commented:
nice feature, if it is implemented, there will be no need worry about `press Enter`.
